polyply/src/apply_links.py

```diff
@@ -247,212 +162,152 @@ def _check_relative_order(resids, orders):
+    nodes and those in the meta_molecule, establish a correspondance
+    between the link nodes and those in the higher resolution
+    molecule of the meta_molecule. Note that the meta_molecule needs
+    to have the "block" attribute discribing which atoms a single
+    meta_molecule node described.
 
-    return graphs, paths
-
-
-def is_subgraph(graph1, graph2):
-    """
-    Checks if graph1 is subgraph isomorphic to graph1.
-
-    Parameters
-    ----------
-    graph1: :class:`networkx.Graph`
-    graph2: :class:`networkx.Graph`
-
-    Returns
-    ----------
-    bool
-    """
-    for node in graph2.nodes:
-        if not node in graph1.nodes:
-           return False
-
-    for edge in graph2.edges:
-        if not graph1.has_edge(edge[0], edge[1]):
-           return False
-
-    return True
-
-
-def _get_link_resnames(link):
-    """
-    Get the resnames of a `link` directly from the node attributes
-    of the link atoms. It is not safe enough to just take the
-    link name, because it is often undefined or empty. Note that
-    a link name can also be a `vermouth.molecule.Choice` object.
-
-    Parameters
-    ----------
-    link: :class:`vermouth.molecule.Link`
+    Parameters:
+    -----------
+    meta_molecule: `:class:polyply.src.meta_molecule.MetaMolecule`
+    link:          `:class:vermouth.molecule.Link`
+    link_to_resid:  dict
+        correspondance dict of link nodes to meta_molecule nodes
 
-    Returns
-    ----------
-    set
-      all unique resnames of a molecule
+    Returns:
+    --------
+    dict
+        correspondance dict of link nodes to atoms in the
+        the meta_molecule.molecule attribute
     """
-    res_names = list(nx.get_node_attributes(link, "resname").values())
-    out_resnames = set()
+    link_to_mol = {}
+    for node in link.nodes:
+        meta_mol_key = link_to_resid[node]
+        block = meta_molecule.nodes[meta_mol_key]["graph"]
+        attrs = link.nodes[node]
+        attrs.update({'resid': meta_mol_key + 1})
+        ignore = ['order', 'charge_group', 'replace']
+        matchs = list(find_atoms(block, ignore=ignore, **attrs))
 
-    for name in res_names:
-        if isinstance(name, vermouth.molecule.Choice):
-            out_resnames.update(name.value)
+        if len(matchs) == 1:
+            link_to_mol[node] = matchs[0]
+        elif len(matchs) == 0:
+            msg = "Found no matchs for node {} in resiue {}. Cannot apply link."
+            raise MatchError(msg.format(node, attrs["resid"]))
         else:
-            out_resnames.add(name)
+            msg = "Found {} matches for node {} in resiue {}. Cannot apply link."
+            raise MatchError(msg.format(len(matchs), node, attrs["resid"]))
 
-    return out_resnames
+    return link_to_mol
 
+def _res_match(node1, node2):
+    ignore = [key for key in node2.keys() if key != "resname"]
+    return attributes_match(node1, node2, ignore_keys=ignore)
```

Comment:
actually what I would prefer is an attributes match method that takes selective arguments that need to match in both of the nodes. The others attributes need to be ignored because:

1) They don't matter and can depend on the input. Remember we also only made the residue graph based on orders and residue names. Now we want to look where these two attributes are solving the LCIS problem on the residue graph. If we wanted to take another argument into account it should already have been considered when making the residue graph. The other attributes of the residue graph are essentially propagated by chance. Sometimes it's the charge group but sometimes it is also the atype, if that is actually by chance the same for all atoms within a residue. 

2) The other attributes at higher resolution level are filtered against in the next step. So we do take into account all attributes that matter at the low resolution level.

Reply:
Also because of the choice object this is not the same as getting the resname attributes